### PR TITLE
Update to RNN documentation (issue #106085)

### PR DESCRIPTION
**Pull request by hasteinmetz**
Addresses [issue #106085](https://github.com/pytorch/pytorch/issues/106085).

In `torch/nn/modules/rnn.py`:
- Adds documentation string to RNNBase class.
- Adds parameters to __init__ methods for RNN, LSTM, and GRU, classes.
- Adds type annotations to __init__ methods for RNN, LSTM, and GRU.

In `torch/ao/nn/quantized/dynamic/modules/rnn.py`:
- Adds type specifications to `_FLOAT_MODULE` attributes in RNNBase, RNN, LSTM, and GRU classes. 
> This resolves a `mypy` assignment error `Incompatible types in assignment (expression has type "Type[LSTM]", base class "RNNBase" defined the type as "Type[RNNBase]")` that seemed to be a result of fully specified type annotations in `torch/nn/modules/rnn.py`).

### CODE REVIEW

**Comment by hasteinmetz** [comment]
Apologies for my absence. I just rebased the branch from `viable/strict` as suggested by pytorch-bot. Hopefully this will pass the checks.

File: torch/ao/nn/quantized/dynamic/modules/rnn.py [hunk 1]
```diff
@@ -61,7 +62,7 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
 
 class RNNBase(torch.nn.Module):
 
-    _FLOAT_MODULE = nn.RNNBase
+    _FLOAT_MODULE: Type[nn.RNNBase] = nn.RNNBase            # Annotation to ensure subclass attributes check against known type
```
Comment (mikaylagawarecki) [inline torch/ao/nn/quantized/dynamic/modules/rnn.py:65]: Could you move the comment to the line above please 
```suggestion
     # Annotation to ensure subclass attributes check against known type
    _FLOAT_MODULE: Type[nn.RNNBase] = nn.RNNBase
```

Separately, to aid my understanding could you please give a minimal example of when this type annotation is helpful?

Comment (hasteinmetz) [reply torch/ao/nn/quantized/dynamic/modules/rnn.py:65]: Thank you for all of your help! 

Locally I would receive a mypy error without the type annotation: `torch/ao/nn/quantized/dynamic/modules/rnn.py:643:21: error: Incompatible types in assignment (expression has type "Type[GRU]", base class "RNNBase" defined the type as "RNNBase")  [assignment]`. Adding the `Type[nn.RNNBase]` type annotation resolves this issue. I'm looking into this now.

Comment (hasteinmetz) [reply torch/ao/nn/quantized/dynamic/modules/rnn.py]: It seems as though the ` _FLOAT_MODULE: Type[nn.RNNBase] = nn.RNNBase` is required to pass the `mypy` checks. The type annotations for GRU and LSTM classes for stylistic consistency. 

It seems as though the additional specifications in the `rnn.py` triggered the `Incompatible types in assignment` errors, but I'm not entirely sure why...

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -453,17 +458,19 @@ class RNN(RNNBase):
         >>> output, hn = rnn(input, h0)
     """
 
-    def __init__(self, *args, **kwargs):
-        if 'proj_size' in kwargs:
-            raise ValueError("proj_size argument is only supported for LSTM, not RNN or GRU")
-        self.nonlinearity = kwargs.pop('nonlinearity', 'tanh')
+    def __init__(self, input_size: int, hidden_size: int, num_layers: int = 1,
+                 non_linearity: str = 'tanh', bias: bool = True, batch_first: bool = False,
```
[the remark this answers is not in the document]
Comment (hasteinmetz) [reply torch/nn/modules/rnn.py]: Done! Thanks!

File: torch/nn/modules/rnn.py [hunk 3]
```diff
@@ -453,17 +458,19 @@ class RNN(RNNBase):
...
+                 dropout: float = 0., bidirectional: bool = False, device=None, dtype=None) -> None:
+        self.nonlinearity = non_linearity
         if self.nonlinearity == 'tanh':
             mode = 'RNN_TANH'
         elif self.nonlinearity == 'relu':
             mode = 'RNN_RELU'
         else:
-            raise ValueError(f"Unknown nonlinearity '{self.nonlinearity}'")
-        super().__init__(mode, *args, **kwargs)
+            raise ValueError(f"Unknown nonlinearity '{self.nonlinearity}'. Select from: 'tanh', 'relu'.")
```
[the remark this answers is not in the document]
Comment (hasteinmetz) [reply torch/nn/modules/rnn.py]: Thanks! Fixed in recent commit.

File: torch/nn/modules/rnn.py [hunk 4]
```diff
@@ -30,6 +30,17 @@ def apply_permutation(tensor: Tensor, permutation: Tensor, dim: int = 1) -> Tens
...
+    r"""Base class for RNN modules (RNN, LSTM, GRU).
+    
+    Implements aspects of RNNs shared by the RNN, LSTM, and GRU classes, such as module initialization
+    and utility methods for parameter storage management.
+
+    .. note::
+        The forward method is not implemented by the RNNBase class.
+
+    .. note::
+        LSTM and GRU classes override some method implemented by RNNBase.
```
[the remark this answers is not in the document]
Comment (hasteinmetz) [reply torch/nn/modules/rnn.py]: Thanks for catching this mistake! Fixed in the most recent commit.

File: torch/nn/modules/rnn.py [hunk 5]
```diff
@@ -453,17 +464,19 @@ class RNN(RNNBase):
         >>> output, hn = rnn(input, h0)
     """
 
-    def __init__(self, *args, **kwargs):
-        if 'proj_size' in kwargs:
-            raise ValueError("proj_size argument is only supported for LSTM, not RNN or GRU")
-        self.nonlinearity = kwargs.pop('nonlinearity', 'tanh')
+    def __init__(self, input_size: int, hidden_size: int, num_layers: int = 1,
```
Comment (mikaylagawarecki) [inline torch/nn/modules/rnn.py:467]: Hey, many apologies for not catching this earlier but unfortunately, this change from `*args, **kwargs` --> the actual parameters might be BC breaking so we cannot make this change. :( 

To make the documentation render properly and fix https://github.com/pytorch/pytorch/issues/103845, we could instead add the parameter list to the first line of the docstring. (e.g. [what was done here](https://www.internalfb.com/code/fbsource/[d14d9248df3cdb4820ac0c1eead86935f6970179]/fbcode/caffe2/torch/nn/functional.py?lines=34)) should make the docs render the argument list correctly.

Would you be willing to make this change?

Further, to get the typing correct you can do define an `@overload` above 

e.g.

```
@overload
def forward(self, input_size: int, hidden_size: int, num_layers: int = 1,. etc.):
      ...

# leave the current forward untouched
def forward(*args, **kwargs):
```

Comment (hasteinmetz) [reply torch/nn/modules/rnn.py:467]: Not a problem! Thank you so much for clarifying! I'll make this change first thing tomorrow morning and mark the conversation as resolved when I'm done. I really appreciate your help :)

File: torch/ao/nn/quantized/dynamic/modules/rnn.py (hunk 1 above)
Comment (hasteinmetz) [reply torch/ao/nn/quantized/dynamic/modules/rnn.py]: This is no longer an issue given the changes to `rnn.py`. I removed the changes to this file.

File: torch/nn/modules/rnn.py (hunk 5 above)
Comment (mikaylagawarecki) [reply torch/nn/modules/rnn.py:491]: Hey, sorry for the delay! 

I think it should be ok to keep the parameter list in one line, appreciate the effort adding the decorator though!

 For example, we definitely break the 80 character limit (e.g. [here](https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L3538)). You can push a commit to see whether CI will error (perhaps we will need to add a noqa or something to the like).

Comment (hasteinmetz) [reply torch/nn/modules/rnn.py:469]: Wonderful! Thank you so much for your help! I made the changes with my most recent commit and I'll keep an eye out for CI to see if it needs a noqa. 

This was my first PR request, so your help and kindness were very much appreciated :)

**Review by mikaylagawarecki — APPROVED** [review]
Thanks for making this change :)

File: torch/nn/modules/rnn.py [hunk 6]
```diff
@@ -351,9 +362,10 @@ def _replicate_for_data_parallel(self):
 
 
 class RNN(RNNBase):
-    r"""Applies a multi-layer Elman RNN with :math:`\tanh` or :math:`\text{ReLU}` non-linearity to an
-    input sequence.
+    r"""__init__(self,input_size,hidden_size,num_layers=1,nonlinearity='tanh',bias=True,batch_first=False,dropout=0.0,bidirectional=False,device=None,dtype=None)
```
[the remark this answers is not in the document]
Comment (hasteinmetz) [reply torch/nn/modules/rnn.py:365]: Oh, awesome! Thank you for the heads up for the future!

**Pull request closed without merging**